File: master_of_plow_CPP_SDK/src-gen/nevonex/can_agmo_steermotor/CAN_AGMO_SteerMotor.cpp
/*
 * nevonex/can_agmo_steermotor/CAN_AGMO_SteerMotor.cpp
 * Copyright (c) Robert Bosch GmbH. All rights reserved.
 */

#include "CAN_AGMO_SteerMotor.hpp"
#include <ecore/EObject.hpp>
#include <ecore/EClass.hpp>
#include "nevonex/can_agmo_steermotor/Can_agmo_steermotorPackage.hpp"
#include <ecorecpp/mapping.hpp>
#include <nevonex-fcal-platform/log/Logger.hpp>
#include <nevonex/fcb/PublishConnectionFactory.hpp>
#include <nevonex/fcb/SubscriberEnum.hpp>
#include <vector>

/*PROTECTED REGION ID(CAN_AGMO_SteerMotor.cpp) START*/
// Please, enable the protected region if you add manually written code.
// To do this, add the keyword ENABLED before START.
/*PROTECTED REGION END*/

using namespace ::nevonex::can_agmo_steermotor;
using namespace ::nevonex::log;

// Default constructor
CAN_AGMO_SteerMotor::CAN_AGMO_SteerMotor() : m_motor_Heartbeat(0), m_motor_Request(0), m_motor_Response_Encoder_Speed(0), m_motor_Response_EncoderCountValue(0)
{
    // Machine does NOT call _initialize() — Provider manages initialization

    m_interfaceDetailMap =
    {
    { 9428, ::nevonex::types::InterfaceDetails("fek/9428") },
    { 9434, ::nevonex::types::InterfaceDetails("fek/9434") },
    { 9435, ::nevonex::types::InterfaceDetails("fek/9435") } };

    /*PROTECTED REGION ID(CAN_AGMO_SteerMotor__CAN_AGMO_SteerMotor) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

CAN_AGMO_SteerMotor::~CAN_AGMO_SteerMotor()
{
    /*PROTECTED REGION ID(CAN_AGMO_SteerMotor__Destructor) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

// References
::nevonex::fcal::MOTOR_HEARTBEAT_T_R_ptr CAN_AGMO_SteerMotor::getMotor_Heartbeat() const
{
    return m_motor_Heartbeat;
}

void CAN_AGMO_SteerMotor::setMotor_Heartbeat(::nevonex::fcal::MOTOR_HEARTBEAT_T_R_ptr _motor_Heartbeat)
{
    ::nevonex::fcal::MOTOR_HEARTBEAT_T_R_ptr _old_motor_Heartbeat = m_motor_Heartbeat;
    m_motor_Heartbeat = _motor_Heartbeat;
    SDK_FCAL_LOG(SeverityLevel::debug) << "CAN_AGMO_SteerMotor property \"motor_Heartbeat\" has changed.";
    notifyPropertyChange("motor_Heartbeat", _old_motor_Heartbeat, m_motor_Heartbeat);
}

::nevonex::fcal::RAW_CAN_DATA_DLC_8_W_ptr CAN_AGMO_SteerMotor::getMotor_Request() const
{
    return m_motor_Request;
}

void CAN_AGMO_SteerMotor::setMotor_Request(::nevonex::fcal::RAW_CAN_DATA_DLC_8_W_ptr _motor_Request)
{
    setMotor_RequestInternal(_motor_Request);

    // Publish the value to FIL layer
    std::vector< ::ecore::EObject_ptr > _publishPayload;
    _publishPayload.push_back(::ecore::as< ::ecore::EObject >(_this()));
    ::nevonex::fcb::PublishConnectionFactory::getInstance()->publish(
            "CAN_AGMO_SteerMotor.motor_Request.pub", "motor_Request",
            _publishPayload, ::nevonex::fcb::SubscriberEnum::FIL);
}

void CAN_AGMO_SteerMotor::setMotor_RequestInternal(::nevonex::fcal::RAW_CAN_DATA_DLC_8_W_ptr _motor_Request)
{
    ::nevonex::fcal::RAW_CAN_DATA_DLC_8_W_ptr _old_motor_Request = m_motor_Request;
    m_motor_Request = _motor_Request;
    SDK_FCAL_LOG(SeverityLevel::debug) << "CAN_AGMO_SteerMotor property \"motor_Request\" has changed.";
    notifyPropertyChange("motor_Request", _old_motor_Request, m_motor_Request);
}

::ecore::EInt CAN_AGMO_SteerMotor::getMotor_Response_Encoder_Speed() const
{
    return m_motor_Response_Encoder_Speed;
}

void CAN_AGMO_SteerMotor::setMotor_Response_Encoder_Speed(::ecore::EInt _motor_Response_Encoder_Speed)
{
    ::ecore::EInt _old_motor_Response_Encoder_Speed = m_motor_Response_Encoder_Speed;
    m_motor_Response_Encoder_Speed = _motor_Response_Encoder_Speed;
    SDK_FCAL_LOG(SeverityLevel::debug) << "CAN_AGMO_SteerMotor property \"motor_Response_Encoder_Speed\" has changed.";
    notifyPropertyChange("motor_Response_Encoder_Speed", _old_motor_Response_Encoder_Speed, m_motor_Response_Encoder_Speed);
}

::ecore::EFloat CAN_AGMO_SteerMotor::getMotor_Response_EncoderCountValue() const
{
    return m_motor_Response_EncoderCountValue;
}

void CAN_AGMO_SteerMotor::setMotor_Response_EncoderCountValue(::ecore::EFloat _motor_Response_EncoderCountValue)
{
    ::ecore::EFloat _old_motor_Response_EncoderCountValue = m_motor_Response_EncoderCountValue;
    m_motor_Response_EncoderCountValue = _motor_Response_EncoderCountValue;
    SDK_FCAL_LOG(SeverityLevel::debug) << "CAN_AGMO_SteerMotor property \"motor_Response_EncoderCountValue\" has changed.";
    notifyPropertyChange("motor_Response_EncoderCountValue", _old_motor_Response_EncoderCountValue, m_motor_Response_EncoderCountValue);
}

// FCAL timestamp/valid
::ecore::ELong CAN_AGMO_SteerMotor::getMotor_Heartbeat_Timestamp()
{
    auto _detail = m_interfaceDetailMap.find(9428);
    if (_detail == m_interfaceDetailMap.end())
        return -1;
    return _detail->second.getTimestamp();
}

::ecore::EBoolean CAN_AGMO_SteerMotor::isMotor_Heartbeat_Valid()
{
    auto _detail = m_interfaceDetailMap.find(9428);
    if (_detail == m_interfaceDetailMap.end())
        return false;
    return _detail->second.isValid();
}

// FD 미생성 API, 호환용 — Publish 전용 인터페이스에는 유효성/타임스탬프가 없다.
::ecore::ELong CAN_AGMO_SteerMotor::getMotor_Request_Timestamp()
{
    return -1;
}

// FD 미생성 API, 호환용 — Publish 전용 인터페이스에는 유효성/타임스탬프가 없다.
::ecore::EBoolean CAN_AGMO_SteerMotor::isMotor_Request_Valid()
{
    return false;
}

::ecore::ELong CAN_AGMO_SteerMotor::getMotor_Response_Encoder_Speed_Timestamp()
{
    auto _detail = m_interfaceDetailMap.find(9434);
    if (_detail == m_interfaceDetailMap.end())
        return -1;
    return _detail->second.getTimestamp();
}

::ecore::EBoolean CAN_AGMO_SteerMotor::isMotor_Response_Encoder_Speed_Valid()
{
    auto _detail = m_interfaceDetailMap.find(9434);
    if (_detail == m_interfaceDetailMap.end())
        return false;
    return _detail->second.isValid();
}

::ecore::ELong CAN_AGMO_SteerMotor::getMotor_Response_EncoderCountValue_Timestamp()
{
    auto _detail = m_interfaceDetailMap.find(9435);
    if (_detail == m_interfaceDetailMap.end())
        return -1;
    return _detail->second.getTimestamp();
}

::ecore::EBoolean CAN_AGMO_SteerMotor::isMotor_Response_EncoderCountValue_Valid()
{
    auto _detail = m_interfaceDetailMap.find(9435);
    if (_detail == m_interfaceDetailMap.end())
        return false;
    return _detail->second.isValid();
}

::ecore::EBoolean CAN_AGMO_SteerMotor::updateInterfaceDetail(int _topicKey, ::ecore::EBoolean _valid, ::ecore::ELong _timestamp)
{
    auto _detail = m_interfaceDetailMap.find(_topicKey);
    if (_detail == m_interfaceDetailMap.end())
    {
        SDK_FCAL_LOG(SeverityLevel::warning) << _topicKey
                << " interface details are not present in the interface detail map[CAN_AGMO_SteerMotor].";
        return false;
    }

    if (_detail->second.getTimestamp() > _timestamp)
    {
        SDK_FCAL_LOG(SeverityLevel::warning)
                << "Old or duplicate message received. Old TS : "
                << _detail->second.getTimestamp() << " new TS : " << _timestamp
                << " feature: CAN_AGMO_SteerMotor topic:" << _topicKey;
        return false;
    }

    _detail->second.setValid(_valid, _timestamp);
    return true;
}

// PropertyChange delegation (diamond inheritance requires explicit override)
void CAN_AGMO_SteerMotor::addPropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::addPropertyChangeListener(_listener);
}

void CAN_AGMO_SteerMotor::removePropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::removePropertyChangeListener(_listener);
}

void CAN_AGMO_SteerMotor::notifyPropertyChange(::ecore::EString const &_name,
        ::ecore::EJavaObject const &_oldValue,
        ::ecore::EJavaObject const &_newValue)
{
    using namespace ::nevonex::types;
    PropertyChange::notifyPropertyChange(_name, _oldValue, _newValue);
}

File: com.bosch.fsp.master_of_plow.gen/src/com/bosch/nevonex/can_agmo_steermotor/impl/CAN_AGMO_SteerMotor.java
/**
Copyright (c) Robert Bosch GmbH. All rights reserved.
*/
package com.bosch.nevonex.can_agmo_steermotor.impl;

import com.bosch.fsp.logger.FCALLogs;
import com.bosch.fsp.logger.LoggerConstants;

import com.bosch.fsp.platform.InterfaceDetails;

import com.bosch.fsp.runtime.feature.IMachine;

import com.bosch.nevonex.common.ITopicObject;
import com.bosch.nevonex.common.impl.TopicObject;

import com.bosch.fsp.runtime.feature.exception.CommunicationException;
import com.bosch.fsp.runtime.feature.exception.InterfaceDisabledException;
import com.bosch.fsp.runtime.feature.exception.MachineWriteException;

import com.bosch.fsp.runtime.util.internal.FeatureConfig;

import com.bosch.nevonex.fcb.IFCALController;
import com.bosch.nevonex.fcb.SubscriberEnum;
import com.bosch.nevonex.fcb.impl.FCALController;

import java.util.ArrayList;

import org.eclipse.emf.ecore.EObject;

import com.bosch.nevonex.fcal.IMOTOR_HEARTBEAT_T_R;

import com.bosch.nevonex.fcal.IRAW_CAN_DATA_DLC_8_W;

import com.bosch.nevonex.can_agmo_steermotor.ICAN_AGMO_SteerMotor;

import com.bosch.nevonex.types.IPropertyChange;
import com.bosch.nevonex.types.impl.TypesPackage;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import java.lang.reflect.InvocationTargetException;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.eclipse.emf.common.util.BasicEList;
import org.eclipse.emf.common.util.EList;

import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.InternalEObject;

import org.eclipse.emf.ecore.util.EDataTypeUniqueEList;

/**
 * An implementation of the model object '<em><b>CAN_AGMO_SteerMotor</b></em>'.
 * @generated
 */
public class CAN_AGMO_SteerMotor extends TopicObject implements ICAN_AGMO_SteerMotor {
	protected EList<PropertyChangeListener> listeners;
	private final IFCALController publishController = new FCALController();
	protected IMOTOR_HEARTBEAT_T_R motor_Heartbeat;
	protected IRAW_CAN_DATA_DLC_8_W motor_Request;
	protected int motor_Response_Encoder_Speed = 0;
	protected float motor_Response_EncoderCountValue = 0.0f;

	/**
	 * @generated
	 */
	protected CAN_AGMO_SteerMotor() {
		super();
		initInterfaceDetails();
	}

	private void initInterfaceDetails() {
		interfaceDetailMap.put("fek/9428", new InterfaceDetails("fek/9428"));
		interfaceDetailMap.put("fek/9434", new InterfaceDetails("fek/9434"));
		interfaceDetailMap.put("fek/9435", new InterfaceDetails("fek/9435"));
		interfaceDetailMap.put("fek/9436", new InterfaceDetails("fek/9436"));
	}

	/**
	 * @generated
	 */
	@Override
	protected EClass eStaticClass() {
		return Can_agmo_steermotorPackage.Literals.CAN_AGMO_STEER_MOTOR;
	}

	/**
	 * @generated
	 */
	public IMOTOR_HEARTBEAT_T_R getMotor_Heartbeat() {
		if (motor_Heartbeat != null && motor_Heartbeat.eIsProxy()) {
			InternalEObject oldMotor_Heartbeat = (InternalEObject) motor_Heartbeat;
			motor_Heartbeat = (IMOTOR_HEARTBEAT_T_R) eResolveProxy(oldMotor_Heartbeat);
			if (motor_Heartbeat != oldMotor_Heartbeat) {
			}
		}
		return motor_Heartbeat;
	}

	/**
	 * @generated
	 */
	public IMOTOR_HEARTBEAT_T_R basicGetMotor_Heartbeat() {
		return motor_Heartbeat;
	}

	/**
	 * @generated
	 */
	public void setMotor_Heartbeat(IMOTOR_HEARTBEAT_T_R newMotor_Heartbeat) {
		IMOTOR_HEARTBEAT_T_R oldMotor_Heartbeat = motor_Heartbeat;
		motor_Heartbeat = newMotor_Heartbeat;
		FCALLogs.getInstance().log
				.debug(LoggerConstants.LOG_SDK_PREFIX + " CAN_AGMO_SteerMotor property \"motor_Heartbeat\" changed from  "
						+ oldMotor_Heartbeat + " to " + newMotor_Heartbeat);
		notifyPropertyChange("motor_Heartbeat", oldMotor_Heartbeat, newMotor_Heartbeat);
	}

	/**
	 * @generated
	 */
	public long getMotor_Heartbeat_Timestamp() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9428");
		if (interfaceDetails != null) {
			return interfaceDetails.getTimestmap();
		}
		return -1;
	}

	/**
	 * @generated
	 */
	public boolean isMotor_Heartbeat_Valid() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9428");
		if (interfaceDetails != null) {
			return interfaceDetails.isValid();
		}
		return false;
	}

	/**
	 * @generated
	 */
	public IRAW_CAN_DATA_DLC_8_W getMotor_Request() {
		if (motor_Request != null && motor_Request.eIsProxy()) {
			InternalEObject oldMotor_Request = (InternalEObject) motor_Request;
			motor_Request = (IRAW_CAN_DATA_DLC_8_W) eResolveProxy(oldMotor_Request);
			if (motor_Request != oldMotor_Request) {
			}
		}
		return motor_Request;
	}

	/**
	 * @generated
	 */
	public IRAW_CAN_DATA_DLC_8_W basicGetMotor_Request() {
		return motor_Request;
	}

	/**
	 * @generated
	 */
	public synchronized void setMotor_Request(IRAW_CAN_DATA_DLC_8_W newMotor_Request) throws MachineWriteException, CommunicationException, InterfaceDisabledException {
		IRAW_CAN_DATA_DLC_8_W oldMotor_Request = motor_Request;
		motor_Request = newMotor_Request;
		FCALLogs.getInstance().log
				.debug(LoggerConstants.LOG_SDK_PREFIX + " CAN_AGMO_SteerMotor property \"motor_Request\" changed from  "
						+ oldMotor_Request + " to " + newMotor_Request);
		notifyPropertyChange("motor_Request", oldMotor_Request, newMotor_Request);
		// Publish the value to FIL layer, if the value is set from application.
		int publishIndex = FeatureConfig.getInstance().getPublishIndex();
		if (!Thread.currentThread().getStackTrace()[publishIndex].getClassName().contains(getClass().getName())) {
			List<EObject> objectValue = new ArrayList<>();
			objectValue.add(this);
			publishController.publish("CAN_AGMO_SteerMotor.motor_Request.pub", "motor_Request", objectValue,
					SubscriberEnum.FIL);
		}
	}

	/**
	 * @generated
	 */
	public long getMotor_Request_Timestamp() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9436");
		if (interfaceDetails != null) {
			return interfaceDetails.getTimestmap();
		}
		return -1;
	}

	/**
	 * @generated
	 */
	public boolean isMotor_Request_Valid() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9436");
		if (interfaceDetails != null) {
			return interfaceDetails.isValid();
		}
		return false;
	}

	public int getMotor_Response_Encoder_Speed() {
		return motor_Response_Encoder_Speed;
	}

	public void setMotor_Response_Encoder_Speed(int newMotor_Response_Encoder_Speed) {
		motor_Response_Encoder_Speed = newMotor_Response_Encoder_Speed;
	}

	public float getMotor_Response_EncoderCountValue() {
		return motor_Response_EncoderCountValue;
	}

	public void setMotor_Response_EncoderCountValue(float newMotor_Response_EncoderCountValue) {
		motor_Response_EncoderCountValue = newMotor_Response_EncoderCountValue;
	}

	/**
	 * @generated
	 */
	public void addPropertyChangeListener(PropertyChangeListener listener) {
		if (listeners == null) {
			listeners = new BasicEList<>();
		}
		listeners.add(listener);
	}

	/**
	 * @generated
	 */
	public void removePropertyChangeListener(PropertyChangeListener listener) {
		if (listeners != null) {
			listeners.remove(listener);
		}
	}

	/**
	 * @generated
	 */
	public void notifyPropertyChange(String name, Object oldValue, Object newValue) {
		if (listeners != null) {
			for (PropertyChangeListener listener : this.listeners) {
				listener.propertyChange(new PropertyChangeEvent(this, name, oldValue, newValue));
			}
		}
	}

	/**
	 * @generated
	 */
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_HEARTBEAT:
			if (resolve)
				return getMotor_Heartbeat();
			return basicGetMotor_Heartbeat();
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_REQUEST:
			if (resolve)
				return getMotor_Request();
			return basicGetMotor_Request();
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_RESPONSE_ENCODER_SPEED:
			return getMotor_Response_Encoder_Speed();
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_RESPONSE_ENCODER_COUNT_VALUE:
			return getMotor_Response_EncoderCountValue();
		}
		return super.eGet(featureID, resolve, coreType);
	}

	/**
	 * @generated
	 */
	@SuppressWarnings("unchecked")
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_HEARTBEAT:
			setMotor_Heartbeat((IMOTOR_HEARTBEAT_T_R) newValue);
			return;
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_REQUEST:
			try {
				setMotor_Request((IRAW_CAN_DATA_DLC_8_W) newValue);
				return;
			} catch (Throwable throwable) {
				throw new RuntimeException(throwable);
			}
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_RESPONSE_ENCODER_SPEED:
			setMotor_Response_Encoder_Speed((Integer) newValue);
			return;
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_RESPONSE_ENCODER_COUNT_VALUE:
			setMotor_Response_EncoderCountValue((Float) newValue);
			return;
		}
		super.eSet(featureID, newValue);
	}

	/**
	 * @generated
	 */
	@Override
	public void eUnset(int featureID) {
		switch (featureID) {
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_HEARTBEAT:
			setMotor_Heartbeat((IMOTOR_HEARTBEAT_T_R) null);
			return;
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_REQUEST:
			try {
				setMotor_Request((IRAW_CAN_DATA_DLC_8_W) null);
				return;
			} catch (Throwable throwable) {
				throw new RuntimeException(throwable);
			}
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_RESPONSE_ENCODER_SPEED:
			setMotor_Response_Encoder_Speed(0);
			return;
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_RESPONSE_ENCODER_COUNT_VALUE:
			setMotor_Response_EncoderCountValue(0.0f);
			return;
		}
		super.eUnset(featureID);
	}

	/**
	 * @generated
	 */
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_HEARTBEAT:
			return motor_Heartbeat != null;
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_REQUEST:
			return motor_Request != null;
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_RESPONSE_ENCODER_SPEED:
			return motor_Response_Encoder_Speed != 0;
		case Can_agmo_steermotorPackage.CAN_AGMO_STEER_MOTOR__MOTOR_RESPONSE_ENCODER_COUNT_VALUE:
			return motor_Response_EncoderCountValue != 0.0f;
		}
		return super.eIsSet(featureID);
	}
} //CAN_AGMO_SteerMotor
